### PR TITLE
feat: add lossy option to `parse_json` function

### LINKED ISSUE #269: Add `lossy` option to `parse_json` function
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Users want to decode incoming bytes as JSON even if they contain invalid UTF-8 data in VRL.

### Attempted Solutions

Not currently possible.

### Proposal

Add an optional `lossy` parameter to `parse_json`. Default this to `true` since we expect this to be the most commonly desired behavior.

### References

- https://github.com/vectordotdev/vector/pull/16324
- https://github.com/vectordotdev/vector/issues/11577
- https://github.com/vectordotdev/vector/issues/10571
- https://github.com/vectordotdev/vector/issues/16406

### Version

vector 0.27.0

### PR DESCRIPTION
**Pull request by jorgehermo9**
Closes #269

I would like to know where the default args values should be documented, as the optional `lossy` argument defaults to true in this case and I didn't found anywhere in the code related to that.

For example, in the `parse_xml` function, this is shown in Vector's docs:
image

But couldn't see how can generate files such as https://github.com/vectordotdev/vector/blob/61b1b187169739870d94b6e709a901398f865aac/website/cue/reference/remap/functions/parse_xml.cue#L3. Or has that documentation been created by hand?

### CODE REVIEW

**Comment by jszwedko** [comment]
Nice, thanks @jorgehermo9 !

I'll give this a look shortly. For documentation, we'll need to update the docs in [the Vector repo](https://github.com/vectordotdev/vector/blob/master/website/cue/reference/remap/functions/parse_json.cue). You can see the [`parse_xml` docs](https://github.com/vectordotdev/vector/blob/master/website/cue/reference/remap/functions/parse_xml.cue) for how the defaults are documented.

**Comment by jszwedko** [comment]
> Or has that documentation been created by hand?

It is currently maintained by hand. We would like to eventually generate it though (ref: https://github.com/vectordotdev/vrl/issues/280).

**Comment by jorgehermo9** [comment]
Thanks!

Opened https://github.com/vectordotdev/vector/pull/21076 in order to address the documentation of this

File: changelog.d/269.feature.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+The `parse_json` function now accepts an optional `lossy` parameter (which defaults to `true`).
+This new parameter allows to control whether the UTF-8 decoding should be lossy or not, replacing
+invalid utf-8 sequences with the Unicode replacement character (U+FFFD) if set to `true` or raising an error
+if set to `false` and an invalid utf-8 sequence is found.
```
Comment (jszwedko) [inline changelog.d/269.feature.md:4]:
```suggestion
The `parse_json` function now accepts an optional `lossy` parameter (which defaults to `true`).

This new parameter allows to control whether the UTF-8 decoding should be lossy or not, replacing
invalid utf-8 sequences with the Unicode replacement character (U+FFFD) if set to `true` or raising an error
if set to `false` and an invalid utf-8 sequence is found.
```

It seems like you were going for two paragraphs here so I split this up so it renders as such.

**Review by jszwedko — APPROVED** [review]
Thanks @jorgehermo9 ! Just one nit about the changelog, otherwise this looks good to me. I think the test names are fine as well as the similarity between the test testing the default value and the test explicitly setting it to `true`.

File: changelog.d/269.feature.md (hunk 1 above)
Comment (jorgehermo9) [reply changelog.d/269.feature.md:4]: Yes, missed that! Thank you!

File: changelog.d/269.feature.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+The `parse_json` function now accepts an optional `lossy` parameter (which defaults to `true`).
+
+This new parameter allows to control whether the UTF-8 decoding should be lossy or not, replacing
+invalid utf-8 sequences with the Unicode replacement character (U+FFFD) if set to `true` or raising an error
```
Comment (jorgehermo9) [inline changelog.d/269.feature.md:4]:
```suggestion
invalid UTF-8 sequences with the Unicode replacement character (U+FFFD) if set to `true` or raising an error
```

**Comment by jorgehermo9** [comment]
Thank you very much for your prompt review @jszwedko !!

**Review by pront — APPROVED** [review]
+1 thank you @jorgehermo9

**Pull request merged**